retry configs: tweak about subsequent connections (#498)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

regard the ECH keys as securely replaced by the server. It SHOULD retry the
handshake with a new transport connection, using the retry configurations
supplied by the server. The retry configurations may only be applied to the
retry connection. The client MUST NOT use retry configurations for connections

> beyond the retry. This avoids introducing pinning concerns or a tracking 

Reviewer: ```suggestion
beyond the retry. This avoids introducing pinning concerns or a tracking
```

regard the ECH keys as securely replaced by the server. It SHOULD retry the
handshake with a new transport connection, using the retry configurations
supplied by the server. The retry configurations may only be applied to the
retry connection. The client MUST only use a retry configuration for subsequent
connections for a short (client-defined) duration and MUST revert to use the

> previously-advertised configurations after that duration has passed.  This

Reviewer 2: Using a duration across the entire client application would introduce exactly the tracking vector the text was written to prevent. This allows information to flow across application contexts if, during that duration, two contexts need to talk to the same origin.

The intent of this text was that you do it just for exactly the one high-level connection attempt (the client application likely has a "connect to server with these parameters" procedure somewhere) and nothing else. That is, you use it _exactly_ once, and within the same application context.

That is, "subsequent connections" refers to the ones after the retry connection.

Author: - The existing uses of "subsequent" are still conflicting.
- I get the tracking issue, but wonder what'd we expect wget to do if asked to crawl a site? It seems wrong to expect it to try/fail/try-again for every URL and better to use a retry_config that worked for the all the requests corresponding to one command line invocation.
- I agree "duration" isn't great though.

Reviewer 2: Yup, agreed that the "subsequent" is unclear. Sorry, I was just trying to explain the intent, so the PR could rephrase it better.

> It seems wrong to expect it to try/fail/try-again for every URL and better to use a retry_config that worked for the all the requests corresponding to one command line invocation.

HTTP connection reuse would smooth it over (which needs to define its tracking vector implications elsewhere) but, yes, there is definitely room to be more permissive here. I wrote this bit originally, and my aim was to write down the simplest thing possible because:

* At the time, we hadn't resolved the multi-CDN issue, so it was unclear how a sticky signal would work. Now we have, so this is possible to resolve, if we want to. (It's only good for those cached DNS entries.)
* I didn't want to go down a huge rabbithole to define how stickiness would interact with the client application's correlation goals, at least not when introducing an error correction case. As we saw from the other discussion and the introduction of a tracking vector in this PR, tracking is a complex domain-specific problem and people tend to get very confused when you try to actually define it. 😄

I'm not opposed to defining a more sticky option, though it's no longer an editorial change and we'd include to define solutions to the above problems. (Given it's a lot more fuss, my inclination at the time was to leave it alone until/unless we learn it's a problem.)

Author: So how about saying something like "clients MUST NOT continue to use an ECHConfig from a set of retry_configs for extended periods - ideally a client would only use such an ECHConfig once and ensure the ensuing TLS session is then used for multiple application messages. Clients might however establish multiple TLS sessions using one such ECHConfig

Author: Oh and btw, it's fair to say part of this isn't editorial and so should just be punted for later.

> retry connection. The client MUST only use a retry configuration for subsequent
> connections for a short (client-defined) duration and MUST revert to use the
> previously-advertised configurations after that duration has passed.  This
> avoids introducing pinning concerns or a tracking vector, should a malicious
> server present client-specific retry configurations in order to identify the
> client in a subsequent ECH handshake.

Reviewer: ```suggestion
retry connection. The client MUST NOT use retry configurations for connections
beyond the retry. This avoids introducing pinning concerns or a tracking 
vector, should a malicious server present client-specific retry configurations
in order to identify the client in a subsequent ECH handshake.
```

How about this proposal which flips the text around somewhat and avoids the possible confusion around "subsequent" connections?

Reviewer: cc @davidben @sftcd
